crates/env/src/api.rs

```diff
@@ -321,50 +323,20 @@ where
 /// - If the instantiation process runs out of gas.
 /// - If given insufficient endowment.
 /// - If the returned account ID failed to decode properly.
-pub fn instantiate_contract<E, Args, Salt, R>(
-    params: &CreateParams<E, Args, Salt, R>,
-) -> Result<ink_primitives::ConstructorResult<E::AccountId>>
-where
-    E: Environment,
-    Args: scale::Encode,
-    Salt: AsRef<[u8]>,
-{
-    <EnvInstance as OnInstance>::on_instance(|instance| {
-        TypedEnvBackend::instantiate_contract::<E, Args, Salt, R>(instance, params)
-    })
-}
-
-/// Attempts to instantiate another contract, returning the instantiation result back to the
-/// caller.
-///
-/// # Note
-///
-/// This is a low level way to instantiate another smart contract.
-///
-/// Prefer to use methods on a `ContractRef` or the [`CreateBuilder`](`crate::call::CreateBuilder`)
-/// through [`build_create`](`crate::call::build_create`) instead.
-///
-/// # Errors
-///
-/// - If the code hash is invalid.
-/// - If the arguments passed to the instantiation process are invalid.
-/// - If the instantiation process traps.
-/// - If the instantiation process runs out of gas.
-/// - If given insufficient endowment.
-/// - If the returned account ID failed to decode properly.
-pub fn instantiate_fallible_contract<E, Args, Salt, R, ContractError>(
-    params: &CreateParams<E, Args, Salt, R>,
+pub fn instantiate_contract<E, Args, Salt, R, ContractRef>(
+    params: &CreateParams<E, Args, Salt, R, ContractRef>,
 ) -> Result<
-    ink_primitives::ConstructorResult<core::result::Result<E::AccountId, ContractError>>,
+    ink_primitives::ConstructorResult<<R as InstantiateResult<R>>::Output<ContractRef>>,
 >
 where
     E: Environment,
     Args: scale::Encode,
     Salt: AsRef<[u8]>,
-    ContractError: scale::Decode,
+    R: InstantiateResult<R>,
+    ContractRef: FromAccountId<E>,
```

Comment:
Why not use the `E::AccountId` type directly?

Reply:
I did start with that, but would have had to implement some `map` capability that does `id` e.g. `AccountId -> AccountId` and `AccountId -> Result<(), AccountId>` and it was slightly tricky. Maybe possible but thought it was better just to call `FromAccountId` in place to get the thing working.